### ISSUE TITLE
no-use-before-define false negative

### ISSUE THREAD

**Issue by BridgeAR** [eslint/eslint#10227]
**Tell us about your environment**

* **ESLint Version:** 4.18.0
* **Node Version:** 10.0.0-pre
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

`no-use-before-define: error`

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
function urgs() {
  for (const value of array) {
    console.log(value);
  }

  const array = [];
}
```

**What did you expect to happen?**

An error should be reported for `array` in the first line of the function.

**What actually happened? Please include the actual, raw output from ESLint.**

Nothing is reported.

**Comment by mysticatea** [comment]
Wow, I can't believe that such bug has remained. :flushed:

**Comment by mysticatea** [comment]
I will work on this if anybody has not started.

**Comment by mysticatea** [comment]
Hmm, this bug is tougher than the first glance. Looks to be by some factors which contain breaking change to fix.

1. `no-use-before-define` rule doesn't check `for` scope entirely.
2. `Linter#getScope()` returns a certain wrong scope if the current node is any of `ForStatement`, `ForInStatement`, and `ForOfStatement`. It should return the scopes which their type is `for`, but it returned the parent scope of the `for` scope. We should fix the `Linter#getScope()` method, but it has big impact, maybe. Fortunately, we are making new major version currently.
3. `eslint-scope` package doesn't seem to handle `ForInStatement`/`ForOfStatement` correctly. `Reference` objects of `ForInStatement#right`/`ForOfStatement#right` are missing. It has been stolen by enigmatic TDZ scope.

So it needs the following steps to fix this bug:

1. Fix `eslint-scope` to remove TDZ scope such as I mentioned in https://github.com/eslint/eslint-scope/pull/33#issuecomment-343646474. This is a breaking change.
2. Fix `Linter#getScope()` to return `for` scope for  `ForStatement`, `ForInStatement`, and `ForOfStatement`. This may be a breaking change.
3. Fix `no-use-before-define` rule to check `for` scopes.
